pageserver/src/tenant/mgr.rs

```diff
@@ -807,28 +1044,35 @@ where
     // tenant-wde cleanup operations may take some time (removing the entire tenant directory), we want to
     // avoid holding the lock for the entire process.
     let tenant = {
-        tenants
+        match tenants
             .write()
             .await
-            .get(&tenant_id)
-            .cloned()
+            .get_slot(&tenant_id)
             .ok_or(TenantStateError::NotFound(tenant_id))?
+        {
+            TenantSlot::Attached(t) => Some(t.clone()),
+            TenantSlot::Secondary => None,
+        }
     };
 
     // allow pageserver shutdown to await for our completion
     let (_guard, progress) = completion::channel();
 
-    // whenever we remove a tenant from memory, we don't want to flush and wait for upload
-    let freeze_and_flush = false;
-
-    // shutdown is sure to transition tenant to stopping, and wait for all tasks to complete, so
-    // that we can continue safely to cleanup.
-    match tenant.shutdown(progress, freeze_and_flush).await {
-        Ok(()) => {}
-        Err(_other) => {
-            // if pageserver shutdown or other detach/ignore is already ongoing, we don't want to
-            // wait for it but return an error right away because these are distinct requests.
-            return Err(TenantStateError::IsStopping(tenant_id));
+    // If the tenant was attached, shut it down gracefully.  For secondary
+    // locations this part is not necessary
+    if let Some(attached_tenant) = tenant {
```

Comment:
1. I prefer exhaustive matches.
2. Wondering whether we should have a `TenantSlot::shutdown` method that does the enum dispatch.
   Eventually you _will_ need to do something for Secondary tenant shutdown, e.g., synchronize with the global task that downloads heatmaps and layers. Better to prepare for that now, so we don't forget any call sites likes this one when we add that.

Reply:
1. Updated the code to use a `match`
2. Yes, we should: I've added it to my TODO list on #5367 as a reminder.  The reason this isn't trivial is that the Tenant::shutdown returns a barrier which one might either wait on or not wait on, depending on the calling context.  I think that barrier stuff might be removable in #5367 because InProgress has its own Notify for blocking on things like a shutdown in progress.